packages/grafana-ui/src/components/Table/TableNG/Cells/TableCellNG.tsx

```diff
@@ -250,6 +129,58 @@ const getStyles = (theme: GrafanaTheme2, isRightAligned: boolean, bgHoverColor?:
   }),
 });
 
+type TableNGCellRenderer = (props: Omit<TableCellNGProps, 'children'>) => ReactNode;
+
+const SPARKLINE_RENDERER: TableNGCellRenderer = (props) => <SparklineCell {...props} />;
+const GAUGE_RENDERER: TableNGCellRenderer = (props) => <BarGaugeCell {...props} />;
+const JSON_RENDERER: TableNGCellRenderer = (props) => <JSONCell {...props} />;
+const AUTO_RENDERER: TableNGCellRenderer = (props) => {
+  const { field } = props;
+  // Handle auto cell type detection
+  if (field.type === FieldType.geo) {
+    return <GeoCell {...props} />;
+  }
+  if (field.type === FieldType.frame) {
+    const firstValue = field.values[0];
+    if (isDataFrame(firstValue) && isTimeSeriesFrame(firstValue)) {
+      return SPARKLINE_RENDERER(props);
+    }
+    return JSON_RENDERER(props);
+  }
+  if (field.type === FieldType.other) {
+    return JSON_RENDERER(props);
+  }
+  return <AutoCell {...props} />;
+};
+
+function isCustomCellOptions(options: TableCellOptions): options is TableCustomCellOptions {
+  return options.type === TableCellDisplayMode.Custom;
+}
+
+export const CELL_RENDERERS: Record<TableCellDisplayMode, TableNGCellRenderer> = {
+  [TableCellDisplayMode.Sparkline]: SPARKLINE_RENDERER,
+  [TableCellDisplayMode.Gauge]: GAUGE_RENDERER,
+  [TableCellDisplayMode.BasicGauge]: GAUGE_RENDERER,
+  [TableCellDisplayMode.GradientGauge]: GAUGE_RENDERER,
+  [TableCellDisplayMode.LcdGauge]: GAUGE_RENDERER,
+  [TableCellDisplayMode.JSONView]: JSON_RENDERER,
+  [TableCellDisplayMode.Image]: (props) => <ImageCell {...props} />,
+  [TableCellDisplayMode.DataLinks]: (props) => <DataLinksCell {...props} />,
+  [TableCellDisplayMode.Actions]: (props) => <ActionsCell {...props} />,
+  [TableCellDisplayMode.Custom]: (props) => {
+    const { cellOptions, field, value, rowIdx, frame } = props;
+    if (!isCustomCellOptions(cellOptions)) {
+      return null; // nonsensical case, but better to typeguard it.
+    }
+    const CustomCellComponent = cellOptions.cellComponent;
+    return <CustomCellComponent field={field} value={value} rowIndex={rowIdx} frame={frame} />;
+  },
+  [TableCellDisplayMode.ColorText]: AUTO_RENDERER,
+  [TableCellDisplayMode.ColorBackground]: AUTO_RENDERER,
+  [TableCellDisplayMode.ColorBackgroundSolid]: AUTO_RENDERER,
+  [TableCellDisplayMode.Auto]: AUTO_RENDERER,
+};
```

Comment:
the nice things about doing this with an object instead of a switch case:
- the lookup is always O(1)
- TypeScript will yell at you in this file if a new TableCellDisplayMode is added but not mapped back here